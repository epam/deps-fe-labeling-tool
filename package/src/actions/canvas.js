import { createAction } from 'redux-actions'

const FEATURE_NAME = 'CANVAS'

const setZoom = createAction(
  `${FEATURE_NAME}/SET_ZOOM`
)

const setRotationAngle = createAction(
  `${FEATURE_NAME}/SET_ROTATION_ANGLE`,
  (page, angle) => ({
    page,
    angle
  })
)

const setScale = createAction(
  `${FEATURE_NAME}/SET_SCALE`
)

const resetDefault = createAction(
  `${FEATURE_NAME}/RESET_DEFAULT`
)

export {
  resetDefault,
  setZoom,
  setScale,
  setRotationAngle
}
