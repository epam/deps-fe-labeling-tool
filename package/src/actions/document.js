import { createAction } from 'redux-actions'

const FEATURE_NAME = 'DOCUMENT'

const storeDocument = createAction(`${FEATURE_NAME}/STORE`)

const resetDefault = createAction(
  `${FEATURE_NAME}/RESET_DEFAULT`
)

export {
  resetDefault,
  storeDocument
}
