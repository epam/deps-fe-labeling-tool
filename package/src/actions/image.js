import { createAction } from 'redux-actions'

const FEATURE_NAME = 'IMAGE'

const setImage = createAction(
  `${FEATURE_NAME}/SET_IMAGE`
)

const resetDefault = createAction(
  `${FEATURE_NAME}/RESET_DEFAULT`
)

export {
  resetDefault,
  setImage
}
