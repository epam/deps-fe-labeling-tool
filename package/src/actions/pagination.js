import { createAction } from 'redux-actions'

const FEATURE_NAME = 'PAGINATION'

const openPage = createAction(`${FEATURE_NAME}/OPEN_PAGE`)

const resetDefault = createAction(
  `${FEATURE_NAME}/RESET_DEFAULT`
)

export {
  resetDefault,
  openPage
}
