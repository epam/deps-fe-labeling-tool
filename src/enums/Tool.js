import { DragIcon } from '@/components/Icons/DragIcon'
import { LabelIcon } from '@/components/Icons/LabelIcon'
import { MergeIcon } from '@/components/Icons/MergeIcon'
import { PointerIcon } from '@/components/Icons/PointerIcon'
import { SplitIcon } from '@/components/Icons/SplitIcon'
import { TableIcon } from '@/components/Icons/TableIcon'

const Tool = {
  DETECT_TABLES: 'detectTables',
  LABEL: 'label',
  TABLE: 'table',
  MERGE: 'merge',
  SPLIT: 'split',
  POINTER: 'pointer',
  GRABBING: 'grabbing'
}

const RESOURCE_TOOL = {
  [Tool.LABEL]: 'Label',
  [Tool.TABLE]: 'Table',
  [Tool.DETECT_TABLES]: 'Detect Tables',
  [Tool.POINTER]: 'Pointer',
  [Tool.GRABBING]: 'Drag'
}

const RESOURCE_TOOL_TOOLTIP = {
  [Tool.DETECT_TABLES]: 'Detect Tables (Shift+D)',
  [Tool.GRABBING]: 'Drag (Hold Alt)',
  [Tool.LABEL]: 'Label (Shift+A)',
  [Tool.MERGE]: 'Merge (Shift+M)',
  [Tool.POINTER]: 'Pointer (Shift+X)',
  [Tool.TABLE]: 'Table (Shift+T)',
  [Tool.SPLIT]: 'Split (Shift+K)'
}

const TOOL_TO_ICON = {
  [Tool.POINTER]: PointerIcon,
  [Tool.GRABBING]: DragIcon,
  [Tool.LABEL]: LabelIcon,
  [Tool.TABLE]: TableIcon,
  [Tool.DETECT_TABLES]: TableIcon,
  [Tool.SPLIT]: SplitIcon,
  [Tool.MERGE]: MergeIcon
}

export {
  RESOURCE_TOOL,
  RESOURCE_TOOL_TOOLTIP,
  TOOL_TO_ICON,
  Tool
}
