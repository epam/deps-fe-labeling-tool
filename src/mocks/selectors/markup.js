import { mockUuid } from '@/mocks/mockUuid'
import { Label, LabelType } from '@/models/Label'
import { Relation } from '@/models/Relation'
import { Table, CellValue } from '@/models/Table'

jest.mock('uuid', () => mockUuid)

const mockLabels = [
  new Label(0.1, 0.2, 0.3, 0.4, 'MOCK_FIELD_1', 0, LabelType.KEY, 'MOCK FIELD KEY 1', {}, 0.8),
  new Label(0.5, 0.6, 0.7, 0.8, 'MOCK_FIELD_1', 0, LabelType.VALUE, 'MOCK FIELD VALUE 1', {}, 0.4),
  new Label(0.09, 0.10, 0.11, 0.12, 'MOCK_FIELD_2', 1, LabelType.KEY, 'MOCK FIELD KEY 2', {}, 0.77),
  new Label(0.13, 0.14, 0.15, 0.16, 'MOCK_FIELD_2', 1, LabelType.VALUE, 'MOCK FIELD VALUE 2', {}, 0.66)
]

const mockTables = [
  new Table(
    [0.5, 0.6, 0.8],
    [0.303, 0.32, 0.336],
    [],
    [new CellValue(0, 0, 'value', 0.8)],
    {},
    'MOCK_FIELD_4'
  )
]

const mockRelations = [
  new Relation(mockLabels[0], mockLabels[1]),
  new Relation(mockLabels[2], mockLabels[3])
]
const mockPage = 1
const mockMarkupState = {
  [mockPage]: {
    labels: mockLabels,
    selectedLabelsIds: [
      mockLabels[0].uid
    ],
    initialLabels: mockLabels,
    tables: mockTables,
    selectedTablesIds: [],
    initialTables: mockTables,
    modifiedObjects: []
  }
}

const markupSelector = jest.fn(() => mockMarkupState)

const modifiedObjectsSelector = jest.fn(() => [])

const pageLabelsSelector = jest.fn(() => mockLabels)

const pageTablesSelector = jest.fn(() => mockTables)

const pageSelectedLabelsSelector = jest.fn(() => [mockLabels[0]])

const pageRelationsSelector = jest.fn(() => mockRelations)

const pageSelectedTablesSelector = jest.fn(() => [])

const pageMarkupSelector = jest.fn(() => ({
  labels: mockLabels,
  tables: mockTables
}))

const pageMarkupStateSelector = jest.fn(() => mockMarkupState[mockPage])

const pageSelectedMarkupObjectsSelector = jest.fn(() => [mockLabels[0]])

const mockMarkupSelectors = {
  markupSelector,
  modifiedObjectsSelector,
  pageLabelsSelector,
  pageTablesSelector,
  pageSelectedLabelsSelector,
  pageSelectedTablesSelector,
  pageRelationsSelector,
  pageMarkupStateSelector,
  pageMarkupSelector,
  pageSelectedMarkupObjectsSelector
}

export {
  mockMarkupSelectors,
  mockTables
}
