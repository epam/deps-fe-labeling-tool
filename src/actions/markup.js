import { batch } from 'react-redux'
import { createAction } from 'redux-actions'
import { ActionCreators } from 'redux-undo'
import { Feature } from '@/enums/Feature'
import { Mode } from '@/enums/Mode'
import { PageMarkup } from '@/models/Markup'
import { Settings } from '@/models/Settings'
import { documentNameSelector } from '@/selectors/document'
import {
  markupSelector,
  pageSelectedLabelsSelector,
  pageSelectedTablesSelector
} from '@/selectors/markup'
import { currentPageSelector } from '@/selectors/pagination'
import * as exportService from '@/services/export'
import { clipboardStorage } from '@/storage'

const FEATURE_NAME = 'MARKUP'

// TODO: #1431
const addLabels = createAction(
  `${FEATURE_NAME}/ADD_LABELS`,
  (page, labels) => ({
    page,
    labels
  })
)

const storeMarkup = createAction(`${FEATURE_NAME}/STORE`)

const storeAssignedMarkup = createAction(`${FEATURE_NAME}/UPDATE`)

const storeImportMarkup = createAction(`${FEATURE_NAME}/STORE_IMPORT`)

const resetDefault = createAction(
  `${FEATURE_NAME}/RESET_DEFAULT`
)

const removeLabels = createAction(
  `${FEATURE_NAME}/REMOVE_LABELS`,
  (page, labels) => ({
    page,
    labels
  })
)

const updateLabels = createAction(
  `${FEATURE_NAME}/UPDATE_LABELS`,
  (page, labels) => ({
    page,
    labels
  })
)

const selectLabels = createAction(
  `${FEATURE_NAME}/SELECT_LABELS`,
  (page, labels) => ({
    page,
    labels
  })
)

const clearSelection = createAction(
  `${FEATURE_NAME}/CLEAR_SELECTION`
)

// TODO: #1431
const addTables = createAction(
  `${FEATURE_NAME}/ADD_TABLES`,
  (page, tables) => ({
    page,
    tables
  })
)

const removeTables = createAction(
  `${FEATURE_NAME}/REMOVE_TABLES`,
  (page, tables) => ({
    page,
    tables
  })
)

const selectTables = createAction(
  `${FEATURE_NAME}/SELECT_TABLES`,
  (page, tables) => ({
    page,
    tables
  })
)

const updateTables = createAction(
  `${FEATURE_NAME}/UPDATE_TABLES`,
  (page, tables) => ({
    page,
    tables
  })
)

const insertCopiedMarkup = createAction(
  `${FEATURE_NAME}/INSERT_COPIED_MARKUP`,
  (page, pageMarkup) => ({
    page,
    pageMarkup
  })
)

const updateInitialMarkup = createAction(
  `${FEATURE_NAME}/UPDATE_INITIAL_MARKUP`
)

const updateAllLabels = createAction(
  `${FEATURE_NAME}/UPDATE_ALL_LABELS`
)

const updateAllTables = createAction(
  `${FEATURE_NAME}/UPDATE_ALL_TABLES`
)

const undo = ActionCreators.undo

const redo = ActionCreators.redo

const reset = (stateIndex = 1) => ActionCreators.jumpToPast(stateIndex)

const copyMarkup = () => async (dispatch, getState) => {
  const state = getState()
  const markupToCopy = new PageMarkup(
    pageSelectedLabelsSelector(state),
    pageSelectedTablesSelector(state)
  )

  await clipboardStorage.write(markupToCopy)
}

const pasteMarkup = () => async (dispatch, getState) => {
  const state = getState()
  const currentPage = currentPageSelector(state)
  const markupToPaste = await clipboardStorage.read()

  if (!PageMarkup.isValid(markupToPaste)) {
    return
  }

  dispatch(insertCopiedMarkup(currentPage, markupToPaste))
}

const importMarkup = () => async (dispatch) => {
  const markup = await exportService.importMarkup()

  dispatch(storeImportMarkup(markup))
}

const exportMarkup = () => (dispatch, getState) => {
  const state = getState()
  const documentName = documentNameSelector(state)
  const documentNameForExport = `${documentName}.json`
  const markup = markupSelector(state)
  exportService.exportMarkup(markup, documentNameForExport)
}

const updatePageMarkup = ({ labels, tables }) => (dispatch, getState) => {
  const state = getState()
  const currentPage = currentPageSelector(state)
  batch(() => {
    labels && dispatch(updateLabels(currentPage, labels))
    tables && dispatch(updateTables(currentPage, tables))
  })
}

const updateLabelsWithSettings = (label) => (dispatch, getState) => {
  const { settings } = getState()

  const multiAssign = (
    settings.mode === Mode.MARKUP &&
    Settings.has(settings, Feature.MULTI_ASSIGN_LABELS)
  )

  dispatch(updateAllLabels({
    label,
    multiAssign
  }))
}

const updateTablesWithSettings = (table) => (dispatch, getState) => {
  const { settings } = getState()

  const multiAssign = (
    settings.mode === Mode.MARKUP &&
    Settings.has(settings, Feature.MULTI_ASSIGN_TABLES)
  )

  dispatch(updateAllTables({
    table,
    multiAssign
  }))
}

export {
  resetDefault,
  addLabels,
  storeMarkup,
  storeAssignedMarkup,
  storeImportMarkup,
  removeLabels,
  updateLabels,
  selectLabels,
  addTables,
  removeTables,
  updateTables,
  selectTables,
  clearSelection,
  redo,
  undo,
  copyMarkup,
  pasteMarkup,
  insertCopiedMarkup,
  reset,
  importMarkup,
  exportMarkup,
  updateAllLabels,
  updateAllTables,
  updateInitialMarkup,
  updatePageMarkup,
  updateLabelsWithSettings,
  updateTablesWithSettings
}
