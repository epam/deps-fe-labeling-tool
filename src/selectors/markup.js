import { createSelector } from 'reselect'
import { LabelType } from '@/models/Label'
import { Relation, getRelatedLabels } from '@/models/Relation'
import { createDeepEqualSelector } from '@/utils/selectors'
import { currentPageSelector } from './pagination'

const rootSelector = (state) => state.markup.present

const markupSelector = createDeepEqualSelector(
  [rootSelector],
  (markupRootState) => {
    const markup = Object.entries(markupRootState).reduce((markup, [page, pageMarkup]) => {
      const { labels, tables } = pageMarkup

      if (labels && labels.length) {
        markup[page] = {
          ...(markup[page] || {}),
          labels
        }
      }

      if (tables && tables.length) {
        markup[page] = {
          ...(markup[page] || {}),
          tables
        }
      }

      return markup
    }, {})

    if (!Object.keys(markup).length) {
      return null
    }
    return markup
  }
)

const modifiedObjectsSelector = createSelector(
  [rootSelector],
  (markupRootState) => {
    const modifiedObjects = Object.entries(markupRootState).reduce((modifiedObjects, [page, pageMarkup]) => {
      if (pageMarkup.modifiedObjects && pageMarkup.modifiedObjects.length) {
        modifiedObjects[page] = [
          ...pageMarkup.modifiedObjects
        ]
      }

      return modifiedObjects
    }, {})

    if (!Object.keys(modifiedObjects).length) {
      return null
    }

    return modifiedObjects
  }
)

const pageMarkupSelector = createSelector(
  [markupSelector, currentPageSelector],
  (markupState, page) => (markupState && markupState[page]) || {}
)

const pageMarkupStateSelector = createSelector(
  [rootSelector, currentPageSelector],
  (markupState, page) => markupState[page]
)

// TODO: #1431
const pageLabelsSelector = createSelector(
  [pageMarkupStateSelector],
  (pageMarkupState) => (pageMarkupState && pageMarkupState.labels) || []
)

const pageSelectedLabelsIdsSelector = createSelector(
  [pageMarkupStateSelector],
  (pageMarkupState) => (pageMarkupState && pageMarkupState.selectedLabelsIds) || []
)

const pageSelectedLabelsSelector = createSelector(
  [pageLabelsSelector, pageSelectedLabelsIdsSelector],
  (labels, selectedIds) => selectedIds.map((uid) => labels.find((obj) => obj.uid === uid))
)

// TODO: #1431
const pageRelationsSelector = createSelector(
  [pageLabelsSelector],
  (labels) => labels.reduce((relations, label) => {
    if (label.type === LabelType.VALUE) {
      return relations
    }

    const relatedLabels = getRelatedLabels(labels, label)
    return [
      ...relations,
      ...relatedLabels.map((rl) => new Relation(label, rl))
    ]
  }, [])
)

// TODO: #1431
const pageTablesSelector = createSelector(
  [pageMarkupStateSelector],
  (pageMarkupState) => (pageMarkupState && pageMarkupState.tables) || []
)

const pageSelectedTablesIdsSelector = createSelector(
  [pageMarkupStateSelector],
  (pageMarkupState) => (pageMarkupState && pageMarkupState.selectedTablesIds) || []
)

const pageSelectedTablesSelector = createSelector(
  [pageTablesSelector, pageSelectedTablesIdsSelector],
  (tables, selectedTablesIds) => selectedTablesIds.map((uid) => tables.find((obj) => obj.uid === uid))
)

const pageSelectedMarkupObjectsSelector = createSelector(
  [pageSelectedLabelsSelector, pageSelectedTablesSelector],
  (labels, tables) => [...labels, ...tables]
)

export {
  markupSelector,
  modifiedObjectsSelector,
  pageLabelsSelector,
  pageSelectedLabelsSelector,
  pageRelationsSelector,
  pageTablesSelector,
  pageSelectedTablesSelector,
  pageMarkupSelector,
  pageMarkupStateSelector,
  pageSelectedMarkupObjectsSelector
}
