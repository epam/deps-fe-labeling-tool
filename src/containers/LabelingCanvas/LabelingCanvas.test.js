import { mockCanvasActions } from '@/mocks/actions/canvas'
import { mockImageActions } from '@/mocks/actions/image'
import { mockMarkupActions } from '@/mocks/actions/markup'
import { mockComponent } from '@/mocks/mockComponent'
import { mockReactRedux } from '@/mocks/mockReactRedux'
import { mockCanvasSelectors } from '@/mocks/selectors/canvas'
import { mockDocumentSelectors } from '@/mocks/selectors/document'
import { mockImageSelectors } from '@/mocks/selectors/image'
import { mockMarkupSelectors } from '@/mocks/selectors/markup'
import { mockPaginationSelectors } from '@/mocks/selectors/pagination'
import { mockSettingsSelectors } from '@/mocks/selectors/settings'
import { mockToolsSelectors } from '@/mocks/selectors/tools'
import React from 'react'
import { shallow } from 'enzyme'
import { setZoom } from '@/actions/canvas'
import { setImage } from '@/actions/image'
import { clearSelection } from '@/actions/markup'
import { CanvasBackground } from '@/components/CanvasBackground'
import { CanvasMouseSelection } from '@/components/CanvasMouseSelection'
import { CanvasObjectsSelection } from '@/components/CanvasObjectsSelection'
import { Tool } from '@/enums/Tool'
import { Label } from '@/models/Label'
import { Rectangle } from '@/models/Rectangle'
import { Table, CellValue } from '@/models/Table'
import { zoomSelector, scaleSelector, pageRotationAngleSelector } from '@/selectors/canvas'
import { pageImageUrlSelector } from '@/selectors/document'
import { imageSelector } from '@/selectors/image'
import { currentPageSelector } from '@/selectors/pagination'
import { settingsSelector } from '@/selectors/settings'
import { selectedToolSelector } from '@/selectors/tools'
import { LabelingCanvas as ConnectedCanvas } from './LabelingCanvas'
import { useLabels } from './useLabels'
import { useTables } from './useTables'

jest.mock('react-redux', () => mockReactRedux)
jest.mock('@/selectors/canvas', () => mockCanvasSelectors)
jest.mock('@/selectors/document', () => mockDocumentSelectors)
jest.mock('@/selectors/pagination', () => mockPaginationSelectors)
jest.mock('@/selectors/settings', () => mockSettingsSelectors)
jest.mock('@/selectors/tools', () => mockToolsSelectors)
jest.mock('@/selectors/image', () => mockImageSelectors)
jest.mock('@/selectors/markup', () => mockMarkupSelectors)
jest.mock('@/actions/canvas', () => mockCanvasActions)
jest.mock('@/actions/markup', () => mockMarkupActions)
jest.mock('@/actions/image', () => mockImageActions)
jest.mock('./DefaultCanvasHotKeys', () => mockComponent('DefaultCanvasHotKeys'))

const mockLabels = [
  new Label(
    0.1,
    0.2,
    0.3,
    0.4,
    'MOCK_FIELD_1',
    0,
    'key',
    'MOCK FIELD KEY 1',
    {},
    0.8
  )
]

const mockTables = [
  new Table(
    [0.5, 0.6, 0.8],
    [0.303, 0.32, 0.336],
    [],
    new CellValue(
      0,
      0,
      'value',
      0.8
    ),
    {},
    'MOCK_FIELD_4',
    undefined,
    '9'
  )
]

jest.mock('./useLabels', () => ({
  useLabels: jest.fn(() => ({
    labels: mockLabels,
    selectedLabels: mockLabels,
    createLabels: jest.fn(),
    selectLabels: jest.fn(),
    updateLabels: jest.fn()
  }))
}))

jest.mock('./useTables', () => ({
  useTables: jest.fn(() => ({
    tables: mockTables,
    selectedTables: mockTables,
    createTables: jest.fn(),
    addTables: jest.fn(),
    selectTables: jest.fn(),
    updateTables: jest.fn()
  }))
}))

const {
  WrappedComponent,
  mapStateToProps,
  mapDispatchToProps
} = ConnectedCanvas

describe('Container: LabelingCanvas', () => {
  const defaultProps = {
    imageUrl: 'MOCK_URL'
  }

  const defaultState = { markup: { present: {} } }

  describe('mapStateToProps', () => {
    let props

    beforeEach(() => {
      props = mapStateToProps(defaultState, defaultProps).props
    })

    it('should pass correct props to the wrapped component', () => {
      expect(props).toMatchSnapshot()
    })

    it('should call zoomSelector with state as argument', () => {
      expect(zoomSelector).toHaveBeenCalledWith(defaultState)
    })

    it('should call pageImageUrlSelector with state as argument', () => {
      expect(pageImageUrlSelector).toHaveBeenCalledWith(defaultState)
    })

    it('should call currentPageSelector with state as argument', () => {
      expect(currentPageSelector).toHaveBeenCalledWith(defaultState)
    })

    it('should call scaleSelector with state as argument', () => {
      expect(scaleSelector).toHaveBeenCalledWith(defaultState)
    })

    it('should call imageSelector with state as argument', () => {
      expect(imageSelector).toHaveBeenCalledWith(defaultState)
    })

    it('should call settingsSelector with state as argument', () => {
      expect(settingsSelector).toHaveBeenCalledWith(defaultState)
    })

    it('should call selectedToolSelector with state as argument', () => {
      expect(selectedToolSelector).toHaveBeenCalledWith(defaultState)
    })

    it('should call pageRotationAngleSelector with state as argument', () => {
      expect(pageRotationAngleSelector).toHaveBeenCalledWith(defaultState)
    })
  })

  describe('mapDispatchToProps', () => {
    const { props, dispatch } = mapDispatchToProps()

    it('should dispatch setZoom action when calling to setZoom prop', () => {
      const MOCK_ZOOM = 2

      props.setZoom(MOCK_ZOOM)

      expect(setZoom).toHaveBeenCalledWith(MOCK_ZOOM)
      expect(dispatch).toHaveBeenCalledWith(setZoom(MOCK_ZOOM))
    })

    it('should dispatch clearSelection action when calling to clearSelection prop', () => {
      props.clearSelection()
      expect(clearSelection).toHaveBeenCalled()
      expect(dispatch).toHaveBeenCalledWith(clearSelection())
    })

    it('should dispatch setImage action when calling to setImage prop', () => {
      props.setImage()
      expect(setImage).toHaveBeenCalled()
      expect(dispatch).toHaveBeenCalledWith(setImage())
    })
  })

  describe('Component', () => {
    let wrapper, defaultProps, selection, hookApi

    selection = {
      x: 2,
      y: 2,
      w: 2,
      h: 2
    }

    const rectangle = new Rectangle(0.001, 0.001, 0.001, 0.001)

    beforeEach(() => {
      defaultProps = {
        ...mapStateToProps().props,
        ...mapDispatchToProps().props
      }
      wrapper = shallow(<WrappedComponent {...defaultProps} />)
    })

    it('should render layout correctly', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should call createLabels once, with correct args when onSelectionEnd is called', () => {
      hookApi = useLabels()
      useLabels.mockImplementationOnce(() => hookApi)
      wrapper = shallow(<WrappedComponent {...defaultProps} />)
      const CanvasMouseSelectionWrapper = wrapper.find(CanvasMouseSelection).at(0)
      CanvasMouseSelectionWrapper.props().onSelectionEnd(selection)
      expect(hookApi.createLabels).nthCalledWith(1, rectangle)
    })

    it('should call createTables once, with correct args when onSelectionEnd is called', () => {
      const selectedTool = Tool.TABLE
      hookApi = useTables()
      useTables.mockImplementationOnce(() => hookApi)
      wrapper = shallow(<WrappedComponent {...defaultProps} selectedTool={selectedTool} />)
      const CanvasMouseSelectionWrapper = wrapper.find(CanvasMouseSelection).at(0)
      CanvasMouseSelectionWrapper.props().onSelectionEnd(selection)
      expect(hookApi.createTables).nthCalledWith(1, rectangle)
    })

    it('should call detectTablesProp once, with correct args when onSelectionEnd is called', () => {
      const selectedTool = Tool.DETECT_TABLES
      wrapper = shallow(<WrappedComponent {...defaultProps} selectedTool={selectedTool} />)
      const CanvasMouseSelectionWrapper = wrapper.find(CanvasMouseSelection).at(0)
      CanvasMouseSelectionWrapper.props().onSelectionEnd(selection)
      expect(defaultProps.detectTablesProp).nthCalledWith(1, rectangle)
    })

    it('should not call detectTablesProp with selectedTool type detectTables when onSelectionEnd is called', () => {
      const selectedTool = Tool.DETECT_TABLES
      selection = {
        x: 0.5,
        y: 0.5,
        w: 0.5,
        h: 0.5
      }
      wrapper = shallow(<WrappedComponent {...defaultProps} selectedTool={selectedTool} />)
      const CanvasMouseSelectionWrapper = wrapper.find(CanvasMouseSelection).at(0)
      CanvasMouseSelectionWrapper.props().onSelectionEnd(selection)
      expect(defaultProps.detectTablesProp).not.toHaveBeenCalled()
    })

    it('should call clearSelection with selectedTool type Pointer when onSelectionClear is called', () => {
      const selectedTool = Tool.POINTER
      wrapper = shallow(<WrappedComponent {...defaultProps} selectedTool={selectedTool} />)
      const CanvasObjectsSelectionWrapper = wrapper.find(CanvasObjectsSelection).at(0)
      CanvasObjectsSelectionWrapper.props().onSelectionClear()
      expect(defaultProps.clearSelection).toHaveBeenCalled()
    })

    it('should not render CanvasBackground if scale is 0', () => {
      defaultProps.scale = 0
      wrapper = shallow(<WrappedComponent {...defaultProps} />)
      expect(wrapper.find(CanvasBackground).exists()).toBe(false)
    })
  })
})
