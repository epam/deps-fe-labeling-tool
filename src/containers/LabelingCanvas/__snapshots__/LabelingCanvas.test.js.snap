// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Container: LabelingCanvas Component should render layout correctly 1`] = `
<Fragment>
  <Canvas
    height={768}
    width={1024}
  >
    <CanvasHOC
      setZoom={[MockFunction]}
      zoom={3}
    />
    <CanvasHOC />
    <CanvasHOC
      color="#F2F6FB"
      image={
        Object {
          "height": 1000,
          "width": 1000,
        }
      }
      rotationAngle={0}
      scale={2}
      setImage={[MockFunction]}
      setScale={[MockFunction]}
      url="mockPageImageUrl.pdf"
    />
    <CanvasHOC
      onSelectionEnd={[Function]}
    />
    <CanvasHOC
      image={
        Object {
          "height": 1000,
          "width": 1000,
        }
      }
      isContentVisible={true}
      key="6"
      label={
        Label {
          "confidence": 0.8,
          "content": "MOCK FIELD KEY 1",
          "fieldCode": "MOCK_FIELD_1",
          "h": 0.4,
          "index": 0,
          "meta": Object {},
          "type": "key",
          "typeName": "label",
          "uid": "6",
          "w": 0.3,
          "x": 0.1,
          "y": 0.2,
        }
      }
      onUpdate={[Function]}
      scale={2}
      selectable={false}
    />
    <CanvasHOC
      image={
        Object {
          "height": 1000,
          "width": 1000,
        }
      }
      key="1 + 2"
      relation={
        Relation {
          "from": Label {
            "confidence": 0.8,
            "content": "MOCK FIELD KEY 1",
            "fieldCode": "MOCK_FIELD_1",
            "h": 0.4,
            "index": 0,
            "meta": Object {},
            "type": "key",
            "typeName": "label",
            "uid": "1",
            "w": 0.3,
            "x": 0.1,
            "y": 0.2,
          },
          "to": Label {
            "confidence": 0.4,
            "content": "MOCK FIELD VALUE 1",
            "fieldCode": "MOCK_FIELD_1",
            "h": 0.8,
            "index": 0,
            "meta": Object {},
            "type": "value",
            "typeName": "label",
            "uid": "2",
            "w": 0.7,
            "x": 0.5,
            "y": 0.6,
          },
          "typeName": "relation",
          "uid": "1 + 2",
        }
      }
      scale={2}
    />
    <CanvasHOC
      image={
        Object {
          "height": 1000,
          "width": 1000,
        }
      }
      key="3 + 4"
      relation={
        Relation {
          "from": Label {
            "confidence": 0.77,
            "content": "MOCK FIELD KEY 2",
            "fieldCode": "MOCK_FIELD_2",
            "h": 0.12,
            "index": 1,
            "meta": Object {},
            "type": "key",
            "typeName": "label",
            "uid": "3",
            "w": 0.11,
            "x": 0.09,
            "y": 0.1,
          },
          "to": Label {
            "confidence": 0.66,
            "content": "MOCK FIELD VALUE 2",
            "fieldCode": "MOCK_FIELD_2",
            "h": 0.16,
            "index": 1,
            "meta": Object {},
            "type": "value",
            "typeName": "label",
            "uid": "4",
            "w": 0.15,
            "x": 0.13,
            "y": 0.14,
          },
          "typeName": "relation",
          "uid": "3 + 4",
        }
      }
      scale={2}
    />
    <CanvasHOC
      image={
        Object {
          "height": 1000,
          "width": 1000,
        }
      }
      key="9"
      onUpdate={[Function]}
      scale={2}
      selectable={false}
      table={
        Table {
          "fieldCode": "MOCK_FIELD_4",
          "index": undefined,
          "merges": Array [],
          "meta": Object {},
          "typeName": "table",
          "uid": "9",
          "values": CellValue {
            "column": 0,
            "confidence": 0.8,
            "row": 0,
            "value": "value",
          },
          "xGuidelines": Array [
            0.5,
            0.6,
            0.8,
          ],
          "yGuidelines": Array [
            0.303,
            0.32,
            0.336,
          ],
        }
      }
    />
  </Canvas>
  <DefaultCanvasHotKeys />
</Fragment>
`;

exports[`Container: LabelingCanvas mapStateToProps should pass correct props to the wrapped component 1`] = `
Object {
  "currentPage": 1,
  "image": Object {
    "height": 1000,
    "width": 1000,
  },
  "imageUrl": "mockPageImageUrl.pdf",
  "relations": Array [
    Relation {
      "from": Label {
        "confidence": 0.8,
        "content": "MOCK FIELD KEY 1",
        "fieldCode": "MOCK_FIELD_1",
        "h": 0.4,
        "index": 0,
        "meta": Object {},
        "type": "key",
        "typeName": "label",
        "uid": "1",
        "w": 0.3,
        "x": 0.1,
        "y": 0.2,
      },
      "to": Label {
        "confidence": 0.4,
        "content": "MOCK FIELD VALUE 1",
        "fieldCode": "MOCK_FIELD_1",
        "h": 0.8,
        "index": 0,
        "meta": Object {},
        "type": "value",
        "typeName": "label",
        "uid": "2",
        "w": 0.7,
        "x": 0.5,
        "y": 0.6,
      },
      "typeName": "relation",
      "uid": "1 + 2",
    },
    Relation {
      "from": Label {
        "confidence": 0.77,
        "content": "MOCK FIELD KEY 2",
        "fieldCode": "MOCK_FIELD_2",
        "h": 0.12,
        "index": 1,
        "meta": Object {},
        "type": "key",
        "typeName": "label",
        "uid": "3",
        "w": 0.11,
        "x": 0.09,
        "y": 0.1,
      },
      "to": Label {
        "confidence": 0.66,
        "content": "MOCK FIELD VALUE 2",
        "fieldCode": "MOCK_FIELD_2",
        "h": 0.16,
        "index": 1,
        "meta": Object {},
        "type": "value",
        "typeName": "label",
        "uid": "4",
        "w": 0.15,
        "x": 0.13,
        "y": 0.14,
      },
      "typeName": "relation",
      "uid": "3 + 4",
    },
  ],
  "rotationAngle": 0,
  "scale": 2,
  "selectedTool": "label",
  "settings": Settings {
    "features": Array [
      SubSettings {
        "code": "autoSave",
        "data": Object {
          "interval": 180000,
        },
      },
      "export",
      "import",
      SubSettings {
        "code": "paging",
        "data": Object {
          "initialPage": 1,
        },
      },
      "language",
      "rotation",
      SubSettings {
        "code": "specialSymbols",
        "data": Array [
          "&",
          "$",
          "£",
          "¥",
          "©",
          "®",
          "¢",
          "%",
          "*",
          "@",
          "µ",
          "±",
          "´",
          "²",
          "~",
          "½",
          "«",
          "»",
          "°",
          "–",
          "—",
          "“",
          "”",
        ],
      },
      "manageFields",
      "multiAssignLabels",
      "multiAssignTables",
    ],
    "mode": "default",
    "panels": Array [
      "TOOLBAR",
      "MARKUP_SIDEBAR",
      "LEFT_SIDEBAR",
    ],
    "tools": Array [
      "detectTables",
      "label",
      "table",
      "merge",
      "split",
      "pointer",
      "grabbing",
    ],
  },
  "zoom": 3,
}
`;
